File: 0530-Minimum-Absolute-Difference-in-BST/java-0530/src/Solution2.java
/// Source : https://leetcode.com/problems/minimum-absolute-difference-in-bst/description/
/// Author : liuyubobobo
/// Time   : 2018-04-30

import java.util.ArrayList;
import java.util.Collections;

/// Using inorder to traverse all nodes
///
/// Time Complexity: O(n)
/// Space Complexity: O(n)
class Solution2 {

    public int getMinimumDifference(TreeNode root) {

        ArrayList<Integer> list = new ArrayList<>();
        inOrder(root, list);

        int res = Integer.MAX_VALUE;
        for(int i = 1 ; i < list.size() ; i ++)
            res = Math.min(res, list.get(i) - list.get(i - 1));
        return res;
    }

    private void inOrder(TreeNode node, ArrayList<Integer> list){

        if(node == null)
            return;

        inOrder(node.left, list);
        list.add(node.val);
        inOrder(node.right, list);
    }
}

File: 0022-Generate-Parentheses/py-0022/Solution2.py
# Source : https://leetcode.com/problems/generate-parentheses/
# Author : penpenps
# Time   : 2019-05-28

from typing import List

# Traversal all combinations with left part length as i and right part as n-1-i 
# Time Complexity: O(2^n)
# Space Complexity: O(n)

class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        ans = []
        for i in range(n):
            left = self.generateParenthesis(i)
            if not left:
                left = [""]
            right = self.generateParenthesis(n-1-i)
            if not right:
                right = [""]
            for l in left:
                for r in right:
                    ans.append("("+l+")"+r)
        return ans

if __name__ == '__main__':
	solution = Solution()
	n = 3
	print(solution.generateParenthesis(n))